### ISSUE TITLE
Promise support on all fields

### ISSUE THREAD

**Issue by eboto** [avantcredit/gql2ts#51]
Hi,

Would be super useful if each field of a record accepted either the type itself or a promise to the type. e.g. given a schema:

```graphql
type Author {
  id: Int!
  firstName: String
  lastName: String
  posts: [Post] # the list of Posts by this author
}

type Post {
  id: Int!
  title: String
  author: Author
  votes: Int
}
```

The generated code currently looks like this: 
```typescript
interface IPost {
  __typename: string;
  id: number;
  title: string | null;
  author: IAuthor | null;
  votes: number | null;
}
```

When using Apollo it would be nice on the server side for the generated `IPost` interface to look more like: 

```typescript
interface IPost {
  ...
  author: Author | null | Promise<Author | null>;
  ...
}
```

What would you think of such a change?

**Comment by brettjurgens** [comment]
sorry, I'm a little confused why it would be a promise. Perhaps I'm missing something.

anyway, in https://github.com/avantcredit/gql2ts/pull/46 i plan to support custom formatting functions, so if you provided something like:
```ts
const formatTypeDeclaration = (name: string, type: string) => `${name}: ${type} | Promise<${type}>;
```
it would definitely be possible

does that work? 

I'm hesitant to wrap everything in a promise. that particularly would not work for my use case of this library and I'm sure the use cases of others

**Comment by eboto** [comment]
Hi Brett,

Your solution for #46 will work great for us!

As for our use case. I'm trying to make objects on the server side that implement these interfaces. Some of the objects that I create can not produce the properties in memory and must make remote calls. e.g. in order to get a post's list of authors I must hit the database.

As such, when I try to implement `Post.authors` on that type, I can not return the actual type and must return a promise to the type, which prevents me from implementing the generated interfaces. If the generated interface additionally supported `Promise<T>` then it would be solved!

You could say "well just create a promise to the whole object", but part of the design is that the call for `authors` should only happen lazily when requested by a particular query.

**Comment by brettjurgens** [comment]
@eboto i just merged #46 and published it on npm as `gql2ts@next`

can you make a file like
```js
// overrides.js
const { DEFAULT_TYPE_PRINTER } = require('@gql2ts/language-typescript');

module.exports = {
  default: { 
    printType: (type, isNonNull) => {
      const declaration = DEFAULT_TYPE_PRINTER(type, isNonNull);
      return `${declaration} | Promise<${declaration}>`;
    }
  }
}
```

and run `gql2ts` as `gql2ts {normal parameters} -e ./overrides.js`

and let me know if it works for your usecase?

**Comment by eboto** [comment]
Hey Brett! Thanks for the change. Came across an error while trying it
```
AssertionError: path must be a string
    at Module.require (module.js:467:3)
    at require (internal/module.js:20:19)
    at run (/app/server/node_modules/gql2ts/dist/index.js:24:28)
    at Object.<anonymous> (/app/server/node_modules/gql2ts/dist/index.js:59:5)
```

I think this PR will fix the particular problem https://github.com/avantcredit/gql2ts/pull/53

**Comment by eboto** [comment]
Now I get this error: 
```
Error: Cannot find module '@gql2ts/language-typescript'
    at Function.Module._resolveFilename (module.js:440:15)
    at Function.Module._load (module.js:388:25)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/app/scripts/schema-gen-overrides.js:1:96)
```

due to the import at the top of `overrides.js`.

This could be because gql2ts is installed locally, not globally. However I would like to keep it that way.

**Comment by eboto** [comment]
OK all fixed! This was the final `overrides` that worked for me:

```javascript
module.exports = {
  printType: (type, isNonNull) => {
    const declaration = isNonNull ? type : `${type} | null`;
    return `${declaration} | Promise<${type}>`;
  }
};
```

**Comment by eboto** [comment]
Giving it a try now for our use case. Thanks for building this. =)

**Comment by brettjurgens** [comment]
👍 thanks for the PR, planning on adding some testing to the CLI

**Comment by brettjurgens** [comment]
@eboto just published 1.0.0, can i close this issue?

**Comment by eboto** [comment]
Yes! Congrats on the big release and thanks for being so responsive & helpful on the issue. Closing it =)
